rfcs/2023-01-06-layout.md

```diff
@@ -0,0 +1,461 @@
+
+#### `Flex` component
+
+A helper component over the `Box` component with `display` set as `Flex`. Arguably, this component seems unneeded without flex specific shorthand properties because the behavior can be achieved with a `display="flex"` prop on `Box` but it improves code readability. Supports same properties.
+
+#### `Grid` component
+
+A helper **web only** component over the `Box` component with `display` set as `grid`. Similar to `Flex` component, it may be argued this component is unneeded with `display="grid"` set on `Box` but it improves code readability. Supports same properties.
+
+#### `Spacer` component
+
+A helper component over the `Box` component primarily aimed at putting space in between components. Useful when the spacing is different between each sibling and can't be solved by gap based properties. `Spacer` can also be helpful for spacing in between custom components which might not support a `sx` prop. Sample usage:
+
+```tsx
+<Box
+  display="flex"
+  flexDirection="column"
+  alignItems="center"
+  marginTop="spacing.1"
+  marginBottom="spacing.2"
+  paddingX="spacing.1"
+>
+  <Button>Button 1</Button>
+  <Spacer size="spacing.1" />
+  <Button>Button 2</Button>
+  <Spacer size="spacing.2" />
+  <Button>Button 3</Button>
+</Box>
+```
+
+Supported properties:
+
+| Prop     | Description                        |
+| -------- | ---------------------------------- |
+| `size`   | Sets both height and width equally |
+| `height` | Sets height                        |
+| `width`  | Sets width                         |
+
+All properties are optional and accepts only spacing tokens. If no prop is provided the spacer is set to `flex: 1`, useful for building layouts where you might want a `Spacer` to fill all available space in between siblings.
+
+#### Responsive styles
+
+In order to ease up building responsive UIs all the style prop values will also support an alternate **object based format** that accepts breakpoints specific values, keeping mobile-first approach (`@media(min-width)`). Sample usage:
+
+```tsx
+<Box
+  display="flex"
+  flexDirection={{ base: 'column', l: 'row' }}
+  alignItems="center"
+  marginTop="spacing.1"
+  marginBottom="spacing.2"
+  paddingX="spacing.1"
+  gap={{ base: 'spacing.1', m: 'spacing.2', l: 'spacing.3' }}
+>
+  <Button>Button 1</Button>
+  <Button>Button 2</Button>
+  <Button>Button 3</Button>
+</Box>
+```
+
+The keys are derived from **breakpoints tokens**. An extra `base` key can be added as a helper to define values for breakpoints below the ones provided, for example `{ base: 'column', l: 'row' }` implies the value is `column` for all breakpoints less than `l`. All wider viewports will inherit largest provided breakpoint value because of `min-width`.
```

Comment:
should we make it more simpler like - we provide `default` as the key which will be applied to every breakpoint and then anything that needs to be overridden can be specified explicitly by the consumers?  I've always found this approach confusing even on other external systems.

Reply:
Gave this more thought. I feel these can cause more confusions compared to the min-width approach.

For example if a layout is supposed to appear horizontally across all sizes greater than `m` people will have to specify the prop for all breakpoints like `l`, `xl`, etc.

Some may also forget to pass the value for `xl` expecting it would keep the value from previous breakpoint but instead it would pick the `default` value and make the layout vertical at `xl`.